platform-sdk/swirlds-jasperdb/src/main/java/com/swirlds/merkledb/collections/AbstractLongList.java

```diff
@@ -0,0 +1,545 @@
+        final int chunkIndex = (int) (newIndex / numLongsPerChunk);
+        final C result = chunkList.get(chunkIndex);
+        if (result == null) {
+            final C newChunk = createChunk();
+            // set new chunk if it's not created yet, if it is - release the chunk immediately
+            // and use the one from the list
+            final C oldChunk = chunkList.compareAndExchange(chunkIndex, null, newChunk);
+            if (oldChunk == null) {
+                return newChunk;
+            } else {
+                releaseChunk(newChunk);
+                return oldChunk;
+            }
+        } else {
+            return result;
+        }
+    }
+
+    /**
+     * Deletes values up to {@code newMinValidIndex}, releases memory chunks reserved for these
+     * values. Note that it takes {@code reservedBufferOffset} into account to decide if a chunk has
+     * to be released.
+     *
+     * @param newMinValidIndex new minimal valid index, left boundary of the list
+     */
+    private void shrinkIfNeeded(final long newMinValidIndex) {
+        if (newMinValidIndex >= size()) {
+            // shrinking is certainly not required
+            return;
+        }
+
+        final int firstValidChunkWithBuffer =
+                (int) Math.max((newMinValidIndex - reservedBufferLength) / numLongsPerChunk, 0);
+        final int firstChunkIndexToDelete = firstValidChunkWithBuffer - 1;
+        for (int i = firstChunkIndexToDelete; i >= 0; i--) {
+            final C chunk = chunkList.get(i);
+            if (chunk != null && chunkList.compareAndSet(i, chunk, null)) {
+                releaseChunk(chunk);
+            }
+        }
+
+        // clean up a chunk with data
+        final int firstChunkWithData = (int) (newMinValidIndex / numLongsPerChunk);
+        final long numberOfElementsToCleanUp = (newMinValidIndex % numLongsPerChunk);
+        partialChunkCleanup(firstChunkWithData, numberOfElementsToCleanUp);
+
+        // clean up chunk(s) reserved for buffer
+        for (int i = firstValidChunkWithBuffer; i < firstChunkWithData; i++) {
+            partialChunkCleanup(i, numLongsPerChunk);
+        }
+    }
+
+    /**
+     * Releases a chunk that is no longer in use. Some implementation may preserve the chunk for further use.
+     * @param chunk chunk to release
+     */
+    protected abstract void releaseChunk(C chunk);
+
+    /**
+     * Zeroes out a part of a chunk
+     * @param chunkIndex index of the chunk to clean up
+     * @param entriesToCleanUp number of entries to clean up in the chunk counting from the left boundary
+     */
+    protected abstract void partialChunkCleanup(final int chunkIndex, final long entriesToCleanUp);
+
+    /**
+     * @return a new chunk
+     */
+    protected abstract C createChunk();
+
+    /**
+     * @return number of memory chunks that this list may have
+     */
+    protected int calculateNumberOfChunks(final long rightBoundary) {
```

Comment:
Missing parameter documentation, so I'm not sure what rightBoundary is.
If it's the maximum count of longs, this is an example calculation.
rightBoundary = 9, numLongsPerChunk = 4
((9-1)/5)+1 == (8/5)+1 == 2  Is that intended?  I would expect to see the result be 3.
((9+1)/5)+1 == (10/5)+1 == 3, which might be correct.
int cast probably isn't helping here (also, should it be long rather than int?)
You might consider `Math.round(Math.ceil((double)rightBoundary/(double)numLongsPerChunk))`, but that might not be quite the performance you need...

Reply:
I'll add the parameter to the java doc. I'll rename the `rightBoundary` to `totalNumberOfElements`. 

>((9-1)/5)+1 == (8/5)+1 == 2 Is that intended?

It should be `(9 - 1) / 4 + 1) == 3`, as intended